Feat: Work for class methods

hrdwdmrbl opened the pull request.

Commit:
Feat: Work for class methods

skabbi commented:
Hi @hrdwdmrbl, would you mind writing some tests also?

hrdwdmrbl commented:
Yeah, if you'll review and accept the PR I can write the tests. 

Can you think of any edge cases? Otherwise my 2 test cases would be approximately:
```
class Foo {
  bar: () => 'hello world'
  foo: () => bar()
}
```
```
class Foo {
  foo: () => bar()
  bar: () => 'hello world'
}
```

skabbi commented:
Here are some courtesy of ChatGPT

**Valid:**
```node
    {
      code: `        
        function highLevelFunction() {
            midLevelFunction();
        }
        
        function midLevelFunction() {
            lowLevelFunction();
        }
        
        function lowLevelFunction() {
            // implementation
        }
      `,
    },
    {
      code: `        
        // Functions with no calls
        function independentFunctionOne() {
            // implementation
        }
        
        function independentFunctionTwo() {
            // implementation
        }
      `,
    },
    {
      code: `        
        // Recursive function
        function recursiveFunction() {
            if (condition) {
                recursiveFunction();
            }
        }
      `,
    },
    {
      code: `        
        // External function calls
        function highLevelFunction() {
            externalFunction();
        }
      `,
    },
    {
      code: `        
        // Nested functions
        function outerFunction() {
          function innerFunction() {
              // implementation
          }
          innerFunction();
        }
      `,
    },
    {
      code: `        
        class ExampleClass {
          highLevelMethod() {
              this.midLevelMethod();
          }
      
          midLevelMethod() {
              this.lowLevelMethod();
          }
      
          lowLevelMethod() {
              // Implementation details
          }
      }
      `,
    },
    {
      code: `        
        class RecursiveClass {
          startProcess() {
              this.processStep(5);
          }
      
          processStep(steps) {
              if (steps > 0) {
                  this.processStep(steps - 1);
              }
          }
      }
      `,
    },
    {
      code: `        
        class StaticInstanceClass {
          static staticMethod() {
              // Static method implementation
          }
      
          instanceMethod() {
              StaticInstanceClass.staticMethod();
          }
        }
      `,
    },
    {
      code: `        
        class StaticInstanceClass {
          static staticMethod1() {
              // Static method implementation
          }
          static staticMethod2() {
            // Static method implementation
          }
      
          instanceMethod1() {
              StaticInstanceClass.staticMethod1();
          }
          instanceMethod2() {
            StaticInstanceClass.staticMethod2();
          }
        }
      `,
    },
    {
      code: `        
        class MixedMethodClass {
          instanceMethod() {
              MixedMethodClass.staticMethod();
          }
      
          static staticMethod() {
              // Some implementation
          }
        }
      `,
    },
```

**Invalid:**
```node
  {
      code: `        
        function lowLevelFunction() {
            // implementation
        }
        
        function highLevelFunction() {
            lowLevelFunction();
        }
      `,
      errors: [{ message: `'lowLevelFunction' ${errMsg}` }]
    },

    {
      code: `        
        class ExampleClass {
          midLevelMethod() {
              this.lowLevelMethod();
          }
      
          highLevelMethod() {
              this.midLevelMethod();
          }
      
          lowLevelMethod() {
              // Implementation details
          }
      }
      `,
      errors: [{ message: `'midLevelMethod' ${errMsg}` }]
    },
    {
      code: `        
        class ExampleClass {
          midLevelMethod() {
              this.lowLevelMethod();
          }
      
          highLevelMethod() {
              this.midLevelMethod();
          }
      
          lowLevelMethod() {
              // Implementation details
          }
      }
      `,
      errors: [{ message: `'midLevelMethod' ${errMsg}` }]
    },
    {
      code: `        
        class ExternalDependencyClass {
          firstMethod() {
              externalFunction();
          }
      
          secondMethod() {
              this.firstMethod();
          }
      }
      `,
      errors: [{ message: `'firstMethod' ${errMsg}` }]
    },
    {
      code: `
        class MixedMethodClass {
          static staticMethod() {
              const instance = new MixedMethodClass();
              instance.instanceMethod();
          }
      
          instanceMethod() {
              // Some implementation
          }
      }
      `,
      errors: [{ message: `'staticMethod' ${errMsg}` }]
    },
    {
      code: `
        class InterleavedMethodClass {
          static firstStaticMethod() {
              // Some implementation
          }
      
          instanceMethod() {
              InterleavedMethodClass.secondStaticMethod();
          }
      
          static secondStaticMethod() {
              // Some implementation
          }
       }
       `,
      errors: [{ message: `'firstStaticMethod' ${errMsg}` }]
    }
```

skabbi commented:
Please add the tests first and make sure the generated cases match your expectations before we merge this to the master branch to prevent unexpected behavior.

You can add these tests to _/tests/index.js_ and then run them with `npm run test` in the root of the project

Commit:
Fix: Was erroneously flagging calls on other objects inside a class

Commit:
Test: Tests to validate the class-based logic

hrdwdmrbl commented:
@skabbi Sorry, I meant that if you agree that you will merge if I do the tests then I will do the tests. :) Anyway, I have added the tests!

I did discover and fix an error in the code too!

Also, the last 2 *invalid* cases are actually valid cases so I moved them. PTAL

skabbi commented on lib/rules/the-step-down-rule.js:
```diff
@@ -7,9 +8,36 @@ module.exports = {
+
+        const calledMethodName = node.callee.property.name;
+        const callerMethodNameIndex = methodStack.indexOf(calledMethodName);
+        if (callerMethodNameIndex === -1 || callerMethodNameIndex >= methodStack.length - 1) {
```
Nit: Would have liked to see a named variable here like you did with `isThisCall`, but in no way a showstopper.

hrdwdmrbl replied:
I agree, that would have been nicer

skabbi reviewed (approved):
One minor nit. But overall, LGTM. Will merge
